SwiftCompilerSources/Sources/Optimizer/DataStructures/Stack.swift

```diff
@@ -129,3 +140,83 @@ struct Stack<Element> : CollectionLikeSequence {
   /// TODO: once we have move-only types, make this a real deinit.
   mutating func deinitialize() { removeAll() }
 }
+
+extension Stack {
+  /// Mark a stack location for future iteration.
+  ///
+  /// TODO: Marker should be ~Escapable.
+  struct Marker {
+    let slab: BridgedPassContext.Slab
+    let index: Int
+  }
+
+  var top: Marker { Marker(slab: lastSlab, index: endIndex) }
+
+  struct Segment : CollectionLikeSequence {
+    let low: Marker
+    let high: Marker
+
+    init(in stack: Stack, low: Marker, high: Marker) {
+      if low.slab.data == nil {
+        assert(low.index == 0, "invalid empty stack marker")
+        // `low == nil` and `high == nil` is a valid empty segment,
+        // even though `assertValid(marker:)` would return false.
+        if high.slab.data != nil {
+          stack.assertValid(marker: high)
+        }
+        self.low = Marker(slab: stack.firstSlab, index: 0)
+        self.high = high
+        return
+      }
+      stack.assertValid(marker: low)
+      stack.assertValid(marker: high)
+      self.low = low
+      self.high = high
+    }
+
+    func makeIterator() -> Stack.Iterator {
+      return Iterator(slab: low.slab, index: low.index,
+                      lastSlab: high.slab, endIndex: high.index)
+    }
+  }
+
+  /// Assert that `marker` is valid based on the current `top`.
+  ///
+  /// This is an assert rather than a query because slabs can reuse
+  /// memory leading to a stale marker that appears valid.
+  func assertValid(marker: Marker) {
+    var currentSlab = lastSlab
+    var currentIndex = endIndex
+    while currentSlab.data != marker.slab.data {
+      assert(currentSlab.data != firstSlab.data, "Invalid stack marker")
+      currentSlab = currentSlab.getPrevious()
+      currentIndex = Stack.slabCapacity
+    }
+    assert(marker.index <= currentIndex, "Invalid stack marker")
+  }
+
+  /// Execute the `body` closure, passing it `self` for further
+  /// mutation of the stack and passing `marker` to mark the stack
+  /// position prior to executing `body`. `marker` must not escape the
+  /// `body` closure.
+  mutating func withMarker<R>(
+    _ body: (inout Stack<Element>, Marker) throws -> R) rethrows -> R {
+    return try body(&self, top)
+  }
+
+  /// Record a stack marker, execute a `body` closure, then execute a
+  /// `handleNewElements` closure with the Segment that contains all
+  /// elements that remain on the stack after being pushed on the
+  /// stack while executing `body`. `body` must push more elements
+  /// than it pops.
+  mutating func withMarker<R>(
```

Comment:
Yes it's much nicer but I'm not convinced we should expose the Stack Index. This API was designed to prevent misuse, which would be likely if we act like Stack can have arbitrary Indices

Reply:
@eeckstein The `Stack.Marker` implementation an extremely insignificant aspect of this PR and trivial to change later. But the `Stack.Index` discussion may take longer. If we really do want to expose stack indices, it's probably best if we talk about it separately and you do it in a follow up PR.

Reply:
okay